src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -165,29 +179,194 @@ class DecisionTree(BaseModel):
+                    leaf_indices.shape[0],
+                    obj.parent,
+                    )
+            new_node.visited = 1
+            nodes.append(new_node)
+            if obj.is_left:
+                obj.parent.left_child = new_node
+            else:
+                obj.parent.right_child = new_node
+        self.leaf_nodes = nodes
+
+    # Assumes that each noted has been visited during fit_new_leaf_nodes
+    def __squash_tree(self) -> None:
+
+        decision_queue = []
+        decision_queue.append(self.root)
+        while len(decision_queue) > 0:
+            cur_node = decision_queue.pop(0)
+            # If we don't have a decision node, just continue
+            if not isinstance(cur_node, DecisionNode):
+                continue
+
+            # If left child was not visited, then squash current node and right
+            # child
+            if (cur_node.left_child is None) or (cur_node.left_child.visited ==
+                                                 0):
+                parent = cur_node.parent
+                # Root node
+                if parent is None:
+                    self.root = cur_node.right_child
+                # if current node is left child
+                elif parent.left_child == cur_node:
+                    # update parent to point to the child that has been visited
+                    # instead
+                    parent.left_child = cur_node.right_child
+                else:
+                    parent.right_child = cur_node.right_child
+
+                cur_node.right_child.parent = parent
+
+                # Only add this squashed child to the queue
+                decision_queue.append(cur_node.right_child)
+
+            # Same for the right
+            elif (cur_node.right_child is None) or (cur_node.right_child.visited
+                                                    == 0):
+                parent = cur_node.parent
+
+                # Root node
+                if parent is None:
+                    self.root = cur_node.right_child
+
+                # if current node is left child
+                elif parent.left_child == cur_node:
+                    # update parent to point to the child that has been visited
+                    # instead
+                    parent.left_child = cur_node.left_child
+                else:
+                    parent.right_child = cur_node.left_child
+
+                cur_node.left_child.parent = parent
+
+                # Only add this squashed child to the queue
+                decision_queue.append(cur_node.left_child)
+            else:
+                # Neither need squashing, add both to the queue
+                decision_queue.append(cur_node.left_child)
+                decision_queue.append(cur_node.right_child)
+
+    def refit_leaf_nodes(self, X: np.ndarray, Y: np.ndarray, sample_weight:
+                         np.ndarray, prediction_indices: np.ndarray, **kwargs):
```

Comment:
I think its better to call "prediction_indices" "sample_indices" here. We can then use the separation into prediction/fitting only when talking about honest splitting.

Reply:
I would have to disagree. In case of bootstrap it would still be prediction_indices. The difference is that for bootstrap prediction_indices is None as bootstrap does not make use of them.